tests/contrib/test_oauth_toolkit.py

```diff
@@ -81,3 +88,39 @@ def test_oauth2_toolkit(no_warnings):
     schema = generator.get_schema(request=None, public=True)
 
     assert_schema(schema, 'tests/contrib/test_oauth_toolkit.yml')
+
+
+@mock.patch(
+    'drf_spectacular.settings.spectacular_settings.OAUTH2_FLOWS',
+    ['implicit']
+)
+@mock.patch(
+    'drf_spectacular.settings.spectacular_settings.OAUTH2_REFRESH_URL',
+    'http://127.0.0.1:8000/o/refresh'
+)
+@mock.patch(
+    'drf_spectacular.settings.spectacular_settings.OAUTH2_AUTHORIZATION_URL',
+    'http://127.0.0.1:8000/o/authorize'
+)
+@mock.patch(
+    'oauth2_provider.settings.oauth2_settings.SCOPES_BACKEND_CLASS',
+    TestScopesBackend,
+)
+@pytest.mark.contrib('oauth2_provider')
+def test_oauth2_toolkit_scopes_backend(no_warnings):
+    router = routers.SimpleRouter()
+    router.register('TokenHasReadWriteScope', TokenHasReadWriteScopeViewset, basename='x')
+
+    urlpatterns = [
+        *router.urls,
+        path('o/', include('oauth2_provider.urls', namespace='oauth2_provider')),
+    ]
+
+    generator = SchemaGenerator(patterns=urlpatterns)
+    schema = generator.get_schema(request=None, public=True)
+
```

Comment:
you need to add `validate_schema(schema)` when manually generating the schema. (it is automatically included for `assert_schema` and `generate_schema`)

i suspect above error would have thrown here then, as `flow['scopes']` should also be a dict.

Reply:
You are right, when I added `validate_schema` it correctly detected my wrong scopes definition. I've fixed both things.